lib/tesla/middleware/telemetry.ex

```diff
@@ -3,30 +3,111 @@ if Code.ensure_loaded?(:telemetry) do
     @behaviour Tesla.Middleware
 
     @moduledoc """
-    Send the request time and meta-information through telemetry.
+    Emits events using the `:telemetry` library to expose instrumentation.
 
     ### Example usage
     ```
     defmodule MyClient do
       use Tesla
 
-      plug Tesla.Middleware.Telemetry
+      plug Tesla.Middleware.Telemetry, prefix: [:my_client]
 
     end
 
-    :telemetry.attach("my-tesla-telemetry", [:tesla, :request], fn event, time, meta, config ->
+    :telemetry.attach("my-tesla-telemetry", [:my_client, :tesla, :request, stop], fn event, measurements, meta, config ->
       # Do something with the event
     end)
     ```
 
+    ### Options
+
+    * `:event_prefix` - a list of atoms to prefix to the telemetry event. Defaults to `[]`
+
+    ## Telemetry Events
+
+    * `[:tesla, :request, :start]` - emitted at the beginning of the request.
+      * Measurement: `%{time: System.monotonic_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :stop]` - emitted at the end of the request.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :error]` - emitted at the end of the request when there is an error.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t, reason: term}`
+
+    * `[:tesla, :request, :exception]` - emitted at the end of the request when an exception is raised.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t, exception: Exception.t, stacktrace: Exception.stacktrace}`
+      * `[:tesla, :request]` - This event is emitted for backwards compatibility only and should be considered deprecated.
+      This event can be disabled by setting `config :tesla, Tesla.Middleware.Telemetry, disable_legacy_event: true` in your config.
+
     Please check the [telemetry](https://hexdocs.pm/telemetry/) for the further usage.
     """
 
+    @disable_legacy_event Application.get_env(:tesla, Tesla.Middleware.Telemetry,
```

Comment:
If we use module prefix we need to add a note to recompile tesla (`mix deps.compile --force tesla` ) when changing this config, otherwise it won't be picked up automatically.

Reply:
Good call.